examples/static-contact-dynamics.py

```diff
@@ -0,0 +1,166 @@
+])
+v0 = np.zeros(model.nv)
+a0 = np.zeros(model.nv)
+
+# 1. GRAVITY TERM
+
+# We compute the gravity terms by using the ID at desired configuration q0, with velocity and acceleration being 0. I.e., ID with a = v = 0
+g_grav = pin.rnea(model, data, q0, v0, a0)
+
+g_bl = g_grav[:6]
+g_j = g_grav[6:]
+
+# 2. FIND CONTACTS
+
+# First, we set the frame for our contacts. We assume the contacts are placed at the following 4 frames and they are 3D
+feet_names = ['FL_FOOT', 'FR_FOOT', 'HL_FOOT', 'HR_FOOT']
+feet_ids = [model.getFrameId(n) for n in feet_names]
+bl_id = model.getFrameId('base_link')
+ncontact = len(feet_names)
+
+# Now, we need to find the contact Jacobians appearing in (1).
+# These are the Jacobians that relate the joint velocity  to the velocity of each feet
+Js__feet_q = [
+    np.copy(pin.computeFrameJacobian(model, data, q0, id, pin.LOCAL))
+    for id in feet_ids
+]
+
+Js__feet_bl = [np.copy(J[:3, :6]) for J in Js__feet_q]
+
+# Notice that we can write the equation above as an horizontal stack of Jacobians transposed and vertical stack of contact forces
+Jc__feet_bl_T = np.zeros([6, 3 * ncontact])
+Jc__feet_bl_T[:, :] = np.vstack(Js__feet_bl).T
+
+# Now I only need to do the pinv to compute the contact forces
+
+ls = np.linalg.pinv(Jc__feet_bl_T) @ g_bl  # This is (3)
+
+# Contact forces at local coordinates (at each foot coordinate)
+ls__f = np.split(ls, ncontact)
+
+pin.framesForwardKinematics(model, data, q0)
+
+# Contact forces at base link frame
+ls__bl = []
+for l__f, foot_id in zip(ls__f, feet_ids):
+    l_sp__f = pin.Force(l__f, np.zeros(3))
+    l_sp__bl = data.oMf[bl_id].actInv(data.oMf[foot_id].act(l_sp__f))
+    ls__bl.append(np.copy(l_sp__bl.vector))
+
+print("\n--- CONTACT FORCES ---")
+for l__f, foot_id, name in zip(ls__bl, feet_ids, feet_names):
+    print("Contact force at foot {} expressed at the BL is: {}".format(
+        name, l__f))
+
+# Notice that if we add all the contact forces are equal to the g_grav
+print("Error between contact forces and gravity at base link: {}".format(
+    np.linalg.norm(g_bl - sum(ls__bl))))
+
+# 3. FIND TAU
+# Find Jc__feet_j
+Js_feet_j = [np.copy(J[:3, 6:]) for J in Js__feet_q]
+
+Jc__feet_j_T = np.zeros([12, 3 * ncontact])
+Jc__feet_j_T[:, :] = np.vstack(Js_feet_j).T
+
+# Apply (5)
+tau = g_j - Jc__feet_j_T @ ls
+
+# 4. CROSS CHECKS
+
+# INVERSE DYNAMICS
+# We can compare this torques with the ones one would obtain when computing the ID considering the external forces in ls
+pin.framesForwardKinematics(model, data, q0)
+
+joint_names = ['FL_KFE', 'FR_KFE', 'HL_KFE', 'HR_KFE']
+joint_ids = [model.getJointId(n) for n in joint_names]
+
+fs_ext = [pin.Force(np.zeros(6)) for _ in range(len(model.joints))]
+for idx, joint in enumerate(model.joints):
+    if joint.id in joint_ids:
+        fext__bl = pin.Force(ls__bl[joint_ids.index(joint.id)])
+        fs_ext[idx] = data.oMi[joint.id].actInv(data.oMf[bl_id].act(fext__bl))
+
+tau_rnea = pin.rnea(model, data, q0, v0, a0, fs_ext)
```

Comment:
It is better if you use `pinocchio.computeGeneralizedGravity`.

Reply:
Both are fine.

Reply:
Thank you for the comments, @cmastalli 

In this line, I think I couldn't use computeGeneralizedGravity, since I was really commuting the ID with external forces, just to make sure that the torques computed manually are the same as the ones computed by the ID method.

Maybe you are referring to the computation done in L72. Since I saw that calling this function is equivalent to a call to RNEA I decided to use RNEA. Since this example aims at being a micro tutorial example, I think that it is nice to realize that you can find the gravitational terms by a simple call to the ID algorithm.

Reply:
Yes, you're correct! I was referring to L72. Apologies for the confusion.